### DISCUSSION TITLE
embedder_main bootstrap scripts are missing some globals

### DISCUSSION THREAD

**Discussion by SnirBroshi** [nodejs/help#4778] [Q&A] [answered]
When using `LoadEnvironment` with a string parameter, e.g.

https://github.com/nodejs/node/blob/2a4452a53af65a13db4efae474162a7dcfd38dd5/test/embedding/embedtest.cc#L76-L82

the script gets access to `require` and `process`, but doesn't have access to other variables accessible to bootstrap code in `lib/internal/`, such as `primordials` and `internalBinding`.

This is the code responsible:

https://github.com/nodejs/node/blob/2a4452a53af65a13db4efae474162a7dcfd38dd5/src/node_builtins.cc#L368-L373
https://github.com/nodejs/node/blob/2a4452a53af65a13db4efae474162a7dcfd38dd5/src/node_builtins.cc#L429-L434

This is a problem when copying code from node's bootstrappers, since the code relies on such globals.
It seems intentional that `embedder_main` scripts are not provided these, but is there a reason? @joyeecheung 

Other than `primordials` and `internalBinding` there is also `module` and `exports` which builtin modules get access to.
It seems that `return`ing a value from the script passed to `LoadEnvironment` returns it to the caller, so I'm not sure if having `module.exports` will do anything.

So should I open a PR to give access to `primordials` and `internalBinding`? And what about `module` and `exports`?

**Comment by SnirBroshi** [comment]
Update: #46825 now we also lost access to `require` internals :(

**Comment by joyeecheung** [comment, marked as answer]
I think the reason was that we do not want to be burdened of maintaining any compatibility of the internals for embedders. It might be okay to add an option to expose them, but not by default and this should be guarded with explicit warning that frequent breakages should be expected.
